File: mycollab-services/src/main/java/com/esofthead/mycollab/module/crm/domain/criteria/ContactSearchCriteria.java
/**
 * This file is part of mycollab-services.
 *
 * mycollab-services is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-services is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-services.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.module.crm.domain.criteria;

import java.util.Arrays;

import com.esofthead.mycollab.common.CountryValueFactory;
import com.esofthead.mycollab.common.i18n.GenericI18Enum;
import com.esofthead.mycollab.core.arguments.NumberSearchField;
import com.esofthead.mycollab.core.arguments.SearchCriteria;
import com.esofthead.mycollab.core.arguments.SetSearchField;
import com.esofthead.mycollab.core.arguments.StringSearchField;
import com.esofthead.mycollab.core.db.query.CompositionStringParam;
import com.esofthead.mycollab.core.db.query.ConcatStringParam;
import com.esofthead.mycollab.core.db.query.DateParam;
import com.esofthead.mycollab.core.db.query.Param;
import com.esofthead.mycollab.core.db.query.PropertyListParam;
import com.esofthead.mycollab.core.db.query.PropertyParam;
import com.esofthead.mycollab.core.db.query.StringListParam;
import com.esofthead.mycollab.core.db.query.StringParam;
import com.esofthead.mycollab.module.crm.CrmDataTypeFactory;
import com.esofthead.mycollab.module.crm.i18n.ContactI18nEnum;

/**
 * 
 * @author MyCollab Ltd.
 * @since 1.0
 * 
 */
public class ContactSearchCriteria extends SearchCriteria {
	private static final long serialVersionUID = 1L;

	public static final Param p_name = new ConcatStringParam("contact-firstname",
			ContactI18nEnum.FORM_NAME, "m_crm_contact", new String[] {
					"firstname", "lastname" });

	public static final Param p_leadsource = new StringListParam(
			"contact-leadsource", ContactI18nEnum.FORM_LEAD_SOURCE,
			"m_crm_contact", "leadSource", Arrays.asList(CrmDataTypeFactory
					.getLeadSourceList()));

	public static final Param p_billingCountry = new StringListParam(
			"contact-billingCountry", ContactI18nEnum.FORM_PRIMARY_COUNTRY,
			"m_crm_contact", "primCountry", Arrays.asList(CountryValueFactory
					.getCountryList()));

	public static final Param p_shippingCountry = new StringListParam(
			"contact-shippingCountry", ContactI18nEnum.FORM_OTHER_COUNTRY,
			"m_crm_contact", "otherCountry", Arrays.asList(CountryValueFactory
					.getCountryList()));

	public static final Param p_anyPhone = new CompositionStringParam(
			"contact-anyPhone",
			ContactI18nEnum.FORM_ANY_PHONE,
			new StringParam[] {
					new StringParam("", null, "m_crm_contact", "officePhone"),
					new StringParam("", null, "m_crm_contact", "mobile"),
					new StringParam("", null, "m_crm_contact", "homePhone"),
					new StringParam("", null, "m_crm_contact", "otherPhone"),
					new StringParam("", null, "m_crm_contact", "fax"),
					new StringParam("", null, "m_crm_contact", "assistantPhone") });

	public static final Param p_anyEmail = new CompositionStringParam(
			"contact-anyEmail", ContactI18nEnum.FORM_ANY_EMAIL,
			new StringParam[] { new StringParam("", null, "m_crm_contact",
					"email") });

	public static final Param p_anyCity = new CompositionStringParam(
			"contact-anyCity", ContactI18nEnum.FORM_ANY_CITY,
			new StringParam[] {
					new StringParam("", null, "m_crm_contact", "primCity"),
					new StringParam("", null, "m_crm_contact", "otherCity") });

	public static final Param p_account = new PropertyParam("contact-account",
			ContactI18nEnum.FORM_ACCOUNTS, "m_crm_contact", "accountId");

	public static final Param p_assignee = new PropertyListParam(
			"contact-assignuser", GenericI18Enum.FORM_ASSIGNEE,
			"m_crm_contact", "assignUser");

	public static final Param p_createdtime = new DateParam("contact-createdtime",
			GenericI18Enum.FORM_CREATED_TIME, "m_crm_contact", "createdTime");

	public static final Param p_lastupdatedtime = new DateParam(
			"contact-lastupdatedtime", GenericI18Enum.FORM_LAST_UPDATED_TIME,
			"m_crm_contact", "lastUpdatedTime");

	private StringSearchField contactName;
	private SetSearchField<String> assignUsers;
	private NumberSearchField accountId;
	private StringSearchField firstname;
	private StringSearchField lastname;
	private StringSearchField anyEmail;
	private StringSearchField anyAddress;
	private StringSearchField anyState;
	private SetSearchField<String> countries;
	private StringSearchField anyPhone;
	private StringSearchField anyCity;
	private StringSearchField anyPostalCode;
	private SetSearchField<String> leadSources;
	private NumberSearchField id;
	private NumberSearchField campaignId;
	private NumberSearchField opportunityId;
	private NumberSearchField caseId;

	public StringSearchField getContactName() {
		return contactName;
	}

	public void setContactName(StringSearchField contactName) {
		this.contactName = contactName;
	}

	public void setAssignUsers(SetSearchField<String> assignUsers) {
		this.assignUsers = assignUsers;
	}

	public SetSearchField<String> getAssignUsers() {
		return assignUsers;
	}

	public NumberSearchField getAccountId() {
		return accountId;
	}

	public void setAccountId(NumberSearchField accountId) {
		this.accountId = accountId;
	}

	public StringSearchField getFirstname() {
		return firstname;
	}

	public void setFirstname(StringSearchField firstname) {
		this.firstname = firstname;
	}

	public StringSearchField getLastname() {
		return lastname;
	}

	public void setLastname(StringSearchField lastname) {
		this.lastname = lastname;
	}

	public StringSearchField getAnyEmail() {
		return anyEmail;
	}

	public void setAnyEmail(StringSearchField anyEmail) {
		this.anyEmail = anyEmail;
	}

	public StringSearchField getAnyAddress() {
		return anyAddress;
	}

	public void setAnyAddress(StringSearchField anyAddress) {
		this.anyAddress = anyAddress;
	}

	public StringSearchField getAnyState() {
		return anyState;
	}

	public void setAnyState(StringSearchField anyState) {
		this.anyState = anyState;
	}

	public void setCountries(SetSearchField<String> countries) {
		this.countries = countries;
	}

	public SetSearchField<String> getCountries() {
		return countries;
	}

	public StringSearchField getAnyPhone() {
		return anyPhone;
	}

	public void setAnyPhone(StringSearchField anyPhone) {
		this.anyPhone = anyPhone;
	}

	public StringSearchField getAnyCity() {
		return anyCity;
	}

	public void setAnyCity(StringSearchField anyCity) {
		this.anyCity = anyCity;
	}

	public StringSearchField getAnyPostalCode() {
		return anyPostalCode;
	}

	public void setAnyPostalCode(StringSearchField anyPostalCode) {
		this.anyPostalCode = anyPostalCode;
	}

	public void setLeadSources(SetSearchField<String> leadSources) {
		this.leadSources = leadSources;
	}

	public SetSearchField<String> getLeadSources() {
		return leadSources;
	}

	public void setId(NumberSearchField id) {
		this.id = id;
	}

	public NumberSearchField getId() {
		return id;
	}

	public NumberSearchField getCampaignId() {
		return campaignId;
	}

	public void setCampaignId(NumberSearchField campaignId) {
		this.campaignId = campaignId;
	}

	public NumberSearchField getOpportunityId() {
		return opportunityId;
	}

	public void setOpportunityId(NumberSearchField opportunityId) {
		this.opportunityId = opportunityId;
	}

	public NumberSearchField getCaseId() {
		return caseId;
	}

	public void setCaseId(NumberSearchField caseId) {
		this.caseId = caseId;
	}
}

File: mycollab-services/src/main/java/com/mycollab/module/crm/domain/criteria/ContactSearchCriteria.kt
/**
 * Copyright © MyCollab
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.mycollab.module.crm.domain.criteria

import com.mycollab.common.CountryValueFactory
import com.mycollab.common.i18n.GenericI18Enum
import com.mycollab.db.arguments.NumberSearchField
import com.mycollab.db.arguments.SearchCriteria
import com.mycollab.db.arguments.SetSearchField
import com.mycollab.db.arguments.StringSearchField
import com.mycollab.db.query.*
import com.mycollab.module.crm.CrmDataTypeFactory
import com.mycollab.module.crm.CrmTypeConstants
import com.mycollab.module.crm.i18n.ContactI18nEnum

import java.util.Arrays

/**
 * @author MyCollab Ltd.
 * @since 1.0
 */
class ContactSearchCriteria : SearchCriteria() {

    var contactName: StringSearchField? = null
    var assignUsers: SetSearchField<String>? = null
    var accountId: NumberSearchField? = null
    var firstname: StringSearchField? = null
    var lastname: StringSearchField? = null
    var anyEmail: StringSearchField? = null
    var anyAddress: StringSearchField? = null
    var anyState: StringSearchField? = null
    var countries: SetSearchField<String>? = null
    var anyPhone: StringSearchField? = null
    var anyCity: StringSearchField? = null
    var anyPostalCode: StringSearchField? = null
    var leadSources: SetSearchField<String>? = null
    var id: NumberSearchField? = null
    var campaignId: NumberSearchField? = null
    var opportunityId: NumberSearchField? = null
    var caseId: NumberSearchField? = null

    companion object {
        private val serialVersionUID = 1L

        @JvmField val p_name: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, GenericI18Enum.FORM_NAME,
                ConcatStringParam("firstname", "m_crm_contact", arrayOf("firstname", "lastname")))

        @JvmField val p_leadsource = CacheParamMapper.register(CrmTypeConstants.CONTACT, ContactI18nEnum.FORM_LEAD_SOURCE,
                I18nStringListParam("leadsource", "m_crm_contact", "leadSource",
                        listOf(*CrmDataTypeFactory.leadSourceList)))

        @JvmField val p_billingCountry: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, ContactI18nEnum.FORM_PRIMARY_COUNTRY,
                StringListParam("billingCountry", "m_crm_contact", "primCountry",
                        Arrays.asList(*CountryValueFactory.countryList!!)))

        @JvmField val p_shippingCountry: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, ContactI18nEnum.FORM_OTHER_COUNTRY,
                StringListParam("shippingCountry", "m_crm_contact", "otherCountry",
                        Arrays.asList(*CountryValueFactory.countryList!!)))

        @JvmField val p_anyPhone: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, ContactI18nEnum.FORM_ANY_PHONE,
                CompositionStringParam("anyPhone",
                        StringParam("", "m_crm_contact", "officePhone"),
                        StringParam("", "m_crm_contact", "mobile"),
                        StringParam("", "m_crm_contact", "homePhone"),
                        StringParam("", "m_crm_contact", "otherPhone"),
                        StringParam("", "m_crm_contact", "fax"),
                        StringParam("", "m_crm_contact", "assistantPhone")))

        @JvmField val p_anyEmail: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, ContactI18nEnum.FORM_ANY_EMAIL,
                CompositionStringParam("anyEmail", StringParam("", "m_crm_contact", "email")))

        @JvmField val p_anyCity: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, ContactI18nEnum.FORM_ANY_CITY,
                CompositionStringParam("anyCity",
                        StringParam("", "m_crm_contact", "primCity"),
                        StringParam("", "m_crm_contact", "otherCity")))

        @JvmField val p_account: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, ContactI18nEnum.FORM_ACCOUNTS,
                PropertyParam("account", "m_crm_contact", "accountId"))

        @JvmField val p_assignee: Param = CacheParamMapper.register<PropertyListParam<*>>(CrmTypeConstants.CONTACT, GenericI18Enum.FORM_ASSIGNEE,
                PropertyListParam<String>("assignuser", "m_crm_contact", "assignUser"))

        @JvmField val p_createdtime: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, GenericI18Enum.FORM_CREATED_TIME,
                DateParam("createdtime", "m_crm_contact", "createdTime"))

        @JvmField val p_lastupdatedtime: Param = CacheParamMapper.register(CrmTypeConstants.CONTACT, GenericI18Enum.FORM_LAST_UPDATED_TIME,
                DateParam("lastupdatedtime", "m_crm_contact", "lastUpdatedTime"))
    }
}
